src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs

```diff
@@ -361,41 +349,62 @@ public string NameFor<TResult>(Expression<Func<TModel, TResult>> expression)
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            var modelExplorer = GetModelExplorer(expression);
+            var modelExpression = GetModelExpression(expression);
             return GenerateTextBox(
-                modelExplorer,
-                GetExpressionName(expression),
-                modelExplorer.Model,
+                modelExpression.ModelExplorer,
+                modelExpression.Name,
+                modelExpression.Model,
                 format,
                 htmlAttributes);
         }
 
+        /// <summary>
+        /// Gets the <see cref="ModelExpression"/> for <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The <see cref="ModelExpression"/>.</returns>
+        public ModelExpression GetModelExpression<TResult>(Expression<Func<TModel, TResult>> expression)
```

Comment:
I'm neutral about making them public - it does let users create `HtmlHelper<T>` extensions or downcast and use this methods without introducing additional test burden